x-pack/test/functional/services/ml/common_data_grid.ts

```diff
@@ -0,0 +1,209 @@
+      return rows;
+    },
+
+    async getDataGridRows(tableSubj: string) {
+      return (await testSubjects.find(tableSubj)).findAllByTestSubject('dataGridRowCell');
+    },
+
+    async assertEuiDataGridNotEmpty(tableSubj: string) {
+      const dataGridRows = await this.getDataGridRows(tableSubj);
+      expect(dataGridRows.length).to.be.greaterThan(
+        0,
+        `EuiDataGrid '${tableSubj}' should have at least one row (got '${dataGridRows.length}')`
+      );
+    },
+
+    async assertEuiDataGridEmpty(tableSubj: string) {
+      const dataGridRows = await this.getDataGridRows(tableSubj);
+      expect(dataGridRows.length).to.eql(
+        0,
+        `EuiDataGrid '${tableSubj}' should be empty (got '${dataGridRows.length} rows')`
+      );
+    },
+
+    async assertEuiDataGridColumnValues(
+      tableSubj: string,
+      column: number,
+      expectedColumnValues: string[]
+    ) {
+      await retry.tryForTime(20 * 1000, async () => {
+        // get a 2D array of rows and cell values
+        // only parse columns up to the one we want to assert
+        const rows = await this.parseEuiDataGrid(tableSubj, column + 1);
+
+        // reduce the rows data to an array of unique values in the specified column
+        const uniqueColumnValues = rows
+          .map((row) => row[column])
+          .flat()
+          .filter((v, i, a) => a.indexOf(v) === i);
+
+        uniqueColumnValues.sort();
+        // check if the returned unique value matches the supplied filter value
+        expect(uniqueColumnValues).to.eql(
+          expectedColumnValues,
+          `Unique EuiDataGrid '${tableSubj}' column values should be '${expectedColumnValues.join()}' (got ${uniqueColumnValues.join()})`
+        );
+      });
+    },
+
+    async assertEuiDataGridColumnValuesNotEmpty(tableSubj: string, column: number) {
+      await retry.tryForTime(20 * 1000, async () => {
+        // get a 2D array of rows and cell values
+        // only parse columns up to the one we want to assert
+        const rows = await this.parseEuiDataGrid(tableSubj, column + 1);
+
+        // reduce the rows data to an array of unique values in the specified column
+        const uniqueColumnValues = rows
+          .map((row) => row[column])
+          .flat()
+          .filter((v, i, a) => a.indexOf(v) === i);
+
+        uniqueColumnValues.forEach((value) => {
+          // check if the returned unique value is not empty
+          expect(value).to.not.eql('');
+        });
+      });
+    },
+
+    async assertColumnSelectPopoverOpenState(tableSubj: string, expectedState: boolean) {
+      const popoverSelector = this.dataGridSelector(tableSubj, 'dataGridColumnSelectorPopover');
+
+      if (expectedState === true) {
+        await testSubjects.existOrFail(popoverSelector, {
+          timeout: 30 * 1000,
+        });
+      } else {
+        await testSubjects.missingOrFail(popoverSelector, {
+          timeout: 30 * 1000,
+        });
```

Comment:
I'd suggest to lower the timeouts to something like 5 seconds because 30 seconds is a lot for a popover to appear / disappear. Also, the wrapping retry in `toggleColumnSelectPopoverState` runs with a 15 seconds timeout, which would never be triggered if this check times out after 30 seconds.

Reply:
Updated here https://github.com/elastic/kibana/pull/125162/commits/c93ad73f619fb78c61f4369dabd5e7390dc5cbdd